cleverhans/loss.py

```diff
@@ -84,20 +109,38 @@ class CrossEntropy(Loss):
   :param model: Model instance, the model on which to apply the loss.
   :param smoothing: float, amount of label smoothing for cross-entropy.
   :param attack: function, given an input x, return an attacked x'.
+  :param pass_y: bool, if True pass y to the attack
   """
-  def __init__(self, model, smoothing=0., attack=None, **kwargs):
+  def __init__(self, model, smoothing=0., attack=None, pass_y=False,
+               adv_coeff=0.5, attack_params=None,
+               **kwargs):
     if smoothing < 0 or smoothing > 1:
       raise ValueError('Smoothing must be in [0, 1]', smoothing)
     self.kwargs = kwargs
     Loss.__init__(self, model, locals(), attack)
     self.smoothing = smoothing
+    self.adv_coeff = adv_coeff
+    self.pass_y = pass_y
+    self.attack_params = attack_params
 
   def fprop(self, x, y, **kwargs):
     kwargs.update(self.kwargs)
     if self.attack is not None:
-      x = x, self.attack(x)
+      attack_params = copy.copy(self.attack_params)
+      if attack_params is None:
+        attack_params = {}
+      if self.pass_y:
+        attack_params['y'] = y
+      x = x, self.attack.generate(x, **attack_params)
+      coeffs = [1. - self.adv_coeff, self.adv_coeff]
+      assert len(coeffs) == len(x)
+      if self.adv_coeff == 1.:
+        x = (x[1],)
+        coeffs = (coeffs[1],)
     else:
       x = tuple([x])
+      coeffs = [1.]
+      assert len(coeffs) == len(x)
```

Comment:
same - seems impossible to violate? is this just to help future implementers?

Reply:
Thanks, removed it